universe/ValueRefs.cpp

```diff
@@ -316,14 +318,14 @@ MeterType NameToMeter(const std::string_view name) {
     for (std::size_t i = 0; i < NAME_BY_METER.size(); i++)
     {
         if (NAME_BY_METER[i] == name)
-            return static_cast<MeterType>(i);
+            return static_cast<MeterType>(i - 1);
```

Comment:
are you sure this works? `i` is `size_t` which will probably underflow to a big number, rather than the -1 you probably want if `name` is an empty string. not sure what happens if you cast that to `MeterType` with underlying type `signed char`.

Reply:
ideally `NAME_BY_METER` would be `constexpr` so you could `static_assert` to make sure the results are as expected

Reply:
Whether the original operation works would probably be implementation depend so no need to risk it. I switched to regular `int`.

Should the `static_assert` check if `i - 1` is within allowed enum range?

Reply:
Tests are failing, _probably_ because array starts with position 0 or some reference pointer(string_view?) mismatch.

Reply:
Looks like it's failing on `METER_POPULATION` which has index 16. Not sure why and unfortunately I don't have a Windows machine to debug locally. Checking other code paths which may result in a `0.0` return in the test.

Reply:
Fixed that part.

Reply:
> Should the `static_assert` check if `i - 1` is within allowed enum range?''

Rather that `NameToMeter("Population") == MeterType::METER_POPULATION`, `NameToMeter("not a meter") == MeterType::INVALID_METER_TYPE`, and similar. Would need to make NameToMeter `constexpr`, which hopefully wouldn't be complicated by it being separately declared in a header.

Reply:
Added static asserts for `NameToMeter`. Still need to figure out the issue with constness in `NameToMeter` and why it can't easily be `constexpr` as well.

Reply:
For `MeterToName` I'm running into the following errors:
```cpp
[225/285] Building CXX object CMakeFiles/freeorion.dir/UI/ObjectListWnd.cpp.o
In file included from /home/amender/projects/freeorion/UI/ObjectListWnd.cpp:25:
/home/amender/projects/freeorion/UI/../universe/ValueRefs.h:452:59: warning: inline function ‘constexpr std::string_view ValueRef::MeterToName(MeterType)’ used but never defined
  452 | [[nodiscard]] FO_COMMON_API constexpr std::string_view    MeterToName(const MeterType meter);
```
It seems like either `MeterType` is not treated as a literal type or I can't force values passed to `MeterToName` to be const-correct.

Reply:
Isn't constexpr MeterType, not string_view ? (because in this case we want to evaluate MeterType and return something to GUI?)